RecoLocalTracker/Phase2TrackerRecHits/src/VectorHitBuilderAlgorithm.cc

```diff
@@ -0,0 +1,559 @@
+#include "Geometry/CommonDetUnit/interface/GeomDet.h"
+#include "Geometry/CommonTopologies/interface/PixelTopology.h"
+#include "DataFormats/TrackerRecHit2D/interface/VectorHit2D.h"
+#include "RecoLocalTracker/Phase2TrackerRecHits/interface/VectorHitBuilderAlgorithm.h"
+#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "RecoLocalTracker/Records/interface/TkPhase2OTCPERecord.h"
+#include "RecoLocalTracker/Phase2TrackerRecHits/interface/Phase2StripCPE.h"
+
+#include "Geometry/CommonTopologies/interface/PixelTopology.h"
+#include "Geometry/CommonDetUnit/interface/GeomDet.h"
+#include "Geometry/TrackerGeometryBuilder/interface/PixelGeomDetUnit.h"
+
+VectorHitBuilderAlgorithm::VectorHitBuilderAlgorithm(const edm::ParameterSet& conf) :
+  cpeTag_(conf.getParameter<edm::ESInputTag>("CPE")),
+  nMaxVHforeachStack(conf.getParameter<int>("maxVectorHitsinaStack")),
+  barrelCut(conf.getParameter< std::vector< double > >("BarrelCut")),
+  endcapCut(conf.getParameter< std::vector< double > >("EndcapCut")),
+  theFitter(new LinearFit())
+{}
+
+void VectorHitBuilderAlgorithm::initialize(const edm::EventSetup& es)
+{
+  //FIXME:ask Vincenzo
+  /*
+  uint32_t tk_cache_id = es.get<TrackerDigiGeometryRecord>().cacheIdentifier();
+  uint32_t c_cache_id = es.get<TkPhase2OTCPERecord>().cacheIdentifier();
+  
+  if(tk_cache_id != tracker_cache_id) {
+  es.get<TrackerDigiGeometryRecord>().get(tracker);
+  tracker_cache_id = tk_cache_id;
+  }
+  if(c_cache_id != cpe_cache_id) {
+  es.get<TkPhase2OTCPERecord>().get(matcherTag, matcher);
+  es.get<TkPhase2OTCPERecord>().get(cpeTag, cpe);
+  cpe_cache_id = c_cache_id;
+  }
+  */
+
+  // get the geometry and topology
+  edm::ESHandle< TrackerGeometry > geomHandle;
+  es.get< TrackerDigiGeometryRecord >().get( geomHandle );
+  initTkGeom(geomHandle);
+
+  edm::ESHandle< TrackerTopology > tTopoHandle;
+  es.get< TrackerTopologyRcd >().get(tTopoHandle);
+  initTkTopo(tTopoHandle);
+
+  // load the cpe via the eventsetup
+  edm::ESHandle<ClusterParameterEstimator<Phase2TrackerCluster1D> > cpeHandle;
+  es.get<TkPhase2OTCPERecord>().get(cpeTag_, cpeHandle);
+  initCpe(cpeHandle.product());
+}
+
+void VectorHitBuilderAlgorithm::initTkGeom(edm::ESHandle< TrackerGeometry > tkGeomHandle){
+  theTkGeom = tkGeomHandle.product();
+}
+void VectorHitBuilderAlgorithm::initTkTopo(edm::ESHandle< TrackerTopology > tkTopoHandle){
+  theTkTopo = tkTopoHandle.product();
+}
+void VectorHitBuilderAlgorithm::initCpe(const ClusterParameterEstimator<Phase2TrackerCluster1D>* cpeProd){
+  cpe = cpeProd;
+}
+
+double VectorHitBuilderAlgorithm::computeParallaxCorrection(const PixelGeomDetUnit*& geomDetUnit_low, const Point3DBase<float, LocalTag>& lPosClu_low,
+                                                            const PixelGeomDetUnit*& geomDetUnit_upp, const Point3DBase<float, LocalTag>& lPosClu_upp){
```

Comment:
No need to pass pointers by reference (i.e. `const PixelGeomDetUnit *geomDetUnit_low` is sufficient).